thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/detection/trigger/DataAvailabilityTaskScheduler.java

```diff
@@ -0,0 +1,222 @@
+    Map<Long, TaskDTO> runningDetection = retrieveRunningDetection();
+    int taskCount = 0;
+    for (DetectionConfigDTO detectionConfig : detection2DatasetMap.keySet()) {
+      try {
+        long detectionConfigId = detectionConfig.getId();
+        if (!runningDetection.containsKey(detectionConfigId)) {
+          if (isAllDatasetUpdated(detectionConfig, detection2DatasetMap.get(detectionConfig), dataset2RefreshTimeMap)) {
+            //TODO: additional check is required if detection is based on aggregated value across multiple data points
+            createDetectionTask(detectionConfig);
+            taskCount++;
+          } else if (isTimeForFallback(detectionConfig)) {
+            LOG.info("Scheduling a task for detection {} due to the fallback mechanism", detectionConfigId);
+            createDetectionTask(detectionConfig);
+            taskCount++;
+          }
+        } else {
+          LOG.info("Skipping creating detection task for detection {} because task {} is not finished.",
+              detectionConfigId, runningDetection.get(detectionConfigId));
+        }
+      } catch (Exception e) {
+        LOG.error("Error in scheduling a detection...", e);
+      }
+    }
+    ThirdeyeMetricsUtil.eventDrivenScheduledTaskCounter.inc(taskCount);
+    LOG.info("Schedule {} tasks in this run...", taskCount);
+
+  }
+
+  public void start() {
+    executorService.scheduleWithFixedDelay(this, 0, delayInSec, TimeUnit.SECONDS);
+  }
+
+  public void close() {
+    executorService.shutdownNow();
+  }
+
+  private void populateDetectionMapAndDataset2RefreshTimeMap(
+      Map<DetectionConfigDTO, Set<String>> dataset2DetectionMap, Map<String, Long> dataset2RefreshTimeMap) {
+    Map<Long, Set<String>> metricCache = new HashMap<>();
+    List<DetectionConfigDTO> detectionConfigs = detectionConfigDAO.findAllActive()
+        .stream().filter(DetectionConfigBean::isDataAvailabilitySchedule).collect(Collectors.toList());
+    for (DetectionConfigDTO detectionConfig : detectionConfigs) {
+      List<String> metricUrns = DetectionConfigFormatter
+          .extractMetricUrnsFromProperties(detectionConfig.getProperties());
+      Set<String> datasets = new HashSet<>();
+      for (String urn : metricUrns) {
+        MetricEntity me = MetricEntity.fromURN(urn);
+        if (!metricCache.containsKey(me.getId())) {
+          datasets.addAll(ThirdEyeUtils.getDatasetConfigsFromMetricUrn(urn)
+              .stream().map(DatasetConfigDTO::getName).collect(Collectors.toList()));
+          // cache the mapping in memory to avoid duplicate retrieval
+          metricCache.put(me.getId(), datasets);
+        } else {
+          // retrieve dataset mapping from memory
+          datasets.addAll(metricCache.get(me.getId()));
+        }
+      }
+      if (datasets.isEmpty()) {
+        LOG.error("No valid dataset is found for detection {}", detectionConfig.getId());
+        continue;
+      }
```

Comment:
Let's not having the error check code everywhere.
When the datasets will be empty?

Reply:
It is possible for the `datasets` to be empty. Since some of our metrics used by active detections are no longer valid, there is no associated dataset with those invalid metrics.